File: python/tests/_executors.py
from __future__ import annotations

from pathlib import Path

from nodes.core.write_plan import DefaultExecutor, WriteOp, WritePlan


class RecordingExecutor:
    """Records every executed plan, then delegates to DefaultExecutor."""

    def __init__(self, root: Path) -> None:
        self.inner = DefaultExecutor(root)
        self.plans: list[list[WriteOp]] = []

    def execute(self, plan: WritePlan) -> None:
        self.plans.append(list(plan))
        self.inner.execute(plan)

File: ts/tests/_executors.ts
import { DefaultExecutor, type WriteOp, type WritePlan, type WritePlanExecutor } from "../src/write-plan.js";

/** Records every executed plan, then delegates to DefaultExecutor. */
export class RecordingExecutor implements WritePlanExecutor {
  readonly inner: DefaultExecutor;
  readonly plans: WriteOp[][] = [];

  constructor(root: string) {
    this.inner = new DefaultExecutor(root);
  }

  execute(plan: WritePlan): void {
    this.plans.push([...plan]);
    this.inner.execute(plan);
  }
}
